src/Storages/MergeTree/ZooKeeperRetries.h

```diff
@@ -0,0 +1,250 @@
+        logLastError("will retry due to error");
+        sleepForMilliseconds(retries_info.curr_backoff_ms);
+        retries_info.curr_backoff_ms = std::min(retries_info.curr_backoff_ms * 2, retries_info.max_backoff_ms);
+
+        /// reset the flag, it will be set to false in case of error
+        iteration_succeeded = true;
+
+        return true;
+    }
+
+    void retryLoop(auto && f)
+    {
+        while (canTry())
+        {
+            try
+            {
+                f();
+            }
+            catch (const zkutil::KeeperException & e)
+            {
+                if (!Coordination::isHardwareError(e.code))
+                    throw;
+
+                setKeeperError(e.code, e.message());
+            }
+        }
+    }
+
+    bool callAndCatchAll(auto && f)
+    {
+        try
+        {
+            f();
+            return true;
+        }
+        catch (const zkutil::KeeperException & e)
+        {
+            setKeeperError(e.code, e.message());
+        }
+        catch (const Exception & e)
+        {
+            setUserError(e.code(), e.what());
+        }
+        return false;
+    }
+
+    void setUserError(int code, std::string message)
+    {
+        if (retries_info.logger)
+            LOG_TRACE(
+                retries_info.logger, "ZooKeeperRetriesControl: {}/{}: setUserError: error={} message={}", retries_info.name, name, code, message);
+
+        /// if current iteration is already failed, keep initial error
+        if (!iteration_succeeded)
+            return;
+
+        iteration_succeeded = false;
+        user_error.code = code;
+        user_error.message = std::move(message);
+        keeper_error = KeeperError{};
+    }
+
+    template <typename... Args>
+    void setUserError(int code, fmt::format_string<Args...> fmt, Args &&... args)
+    {
+        setUserError(code, fmt::format(fmt, std::forward<Args>(args)...));
+    }
+
+    void setKeeperError(Coordination::Error code, std::string message)
+    {
+        if (retries_info.logger)
+            LOG_TRACE(
+                retries_info.logger, "ZooKeeperRetriesControl: {}/{}: setKeeperError: error={} message={}", retries_info.name, name, code, message);
+
+        /// if current iteration is already failed, keep initial error
+        if (!iteration_succeeded)
+            return;
+
+        iteration_succeeded = false;
+        keeper_error.code = code;
+        keeper_error.message = std::move(message);
+        user_error = UserError{};
+    }
+
+    void stopRetries() { stop_retries = true; }
+
+    void requestUnconditionalRetry() { unconditional_retry = true; }
+
+    bool isLastRetry() const { return retries_info.retry_count >= retries_info.max_retries; }
+
+    bool isRetry() const { return retries_info.retry_count > 0; }
+
+    Coordination::Error getLastKeeperErrorCode() const { return keeper_error.code; }
+
+    void actionAfterLastFailedRetry(std::function<void()> f) { action_after_last_failed_retry = std::move(f); }
```

Comment:
Should we save (and call) all lambdas that were ever passed to this function? Should we call them if the last retry succeeded? Or this function can be called only once? Imagine that we are unlucky and on each iteration we get zk error on attempt to commit part to zk. The lambda passed to this function captures `part_name`. If we try to insert part with new name on next iteration, then we will forget to call `enqueuePartForCheck` for previously inserted parts. Seems like it's not possible, because after `actionAfterLastFailedRetry` we will either return or throw in the beginning of the next iteration (and will never try to allocate new block number), but it's not clear from the code and the interface. Consider adding some assertions or moving code that checks if part was committed to zk to the branch that handles connection loss on commit.

Reply:
Added comments in #42607 [c53b96a](https://github.com/ClickHouse/ClickHouse/pull/42607/commits/c53b96a11ce776f96e7cbd3b8cfe74f9fd495526)